### PR TITLE
Block deploys to preview channels if would break prod

### PR DESCRIPTION
**Pull request by inlined**
Thanks to @jamesdaniels clever idea, we are able to dramatically cut scope by erroring if deploying to a preview would break prod rather than simply trying to modify prod while deploying to a preview.

This is an alternative approach to #5717 that is intended to be more surgical.

### CODE REVIEW

File: src/deploy/hosting/prepare.ts [hunk 1]
```diff
@@ -52,3 +61,70 @@ export async function prepare(context: Context, options: HostingOptions & Option
...
+      // This function is new. It can't be pinned elsewhere
+      if (!endpoint) {
+        continue;
+      }
+      targetTaggedRewrites[target] = `${region}/${endpoint.runServiceId || endpoint.id}`;
+    }
+  }
+
+  if (!Object.keys(targetTaggedRewrites).length) {
+    console.error("No targets with tags");
```
Comment (jamesdaniels) [inline src/deploy/hosting/prepare.ts:111]: No need to error right?

**Review by jamesdaniels — APPROVED** [review]
LGTM, doing some verification now.

File: src/deploy/index.ts [hunk 2]
```diff
@@ -67,11 +67,6 @@ export const deploy = async function (
       await prepareFrameworks(targetNames, context, options);
       const nowTargetsFunctions = targetNames.includes("functions");
       if (nowTargetsFunctions && !usedToTargetFunctions) {
-        if (context.hostingChannel && !experiments.isEnabled("pintags")) {
```
Comment (jamesdaniels) [inline src/deploy/index.ts:70]: Let's revert this change for now and capture enabling this experiment by default in a follow on, SG?

File: src/deploy/hosting/prepare.ts [hunk 3]
```diff
@@ -52,3 +61,69 @@ export async function prepare(context: Context, options: HostingOptions & Option
...
+    return source.source;
+  } else if ("regex" in source) {
+    return source.regex;
+  } else {
+    assertExhaustive(source);
+  }
+}
+
+/**
+ *
```
Comment (taeold) [inline src/deploy/hosting/prepare.ts:78]: do you mind adding a short summary of why certain pins are considered unsafe?
